tsdb/shard.go

```diff
@@ -1302,6 +1306,145 @@ func (a Shards) FieldKeysByPredicate(expr influxql.Expr) (map[string][]string, e
 	return all, nil
 }
 
+// consecutiveAndChildren finds all child nodes of consecutive
+// influxql.BinaryExpr with AND operator nodes ("AND nodes") which are not
+// themselves AND nodes. This may be the root of the tree if the root of the
+// tree is not an AND node.
+type consecutiveAndChildren struct {
+	children []influxql.Node
+}
+
+func (v *consecutiveAndChildren) Visit(node influxql.Node) influxql.Visitor {
+	switch n := node.(type) {
+	case *influxql.BinaryExpr:
+		if n.Op == influxql.AND {
+			return v
+		}
+	case *influxql.ParenExpr:
+		// Parens are essentially a no-op and can be traversed through.
+		return v
+	}
+
+	// If this wasn't a BinaryExpr with an AND operator or a Paren, record this
+	// child node and stop the search for this branch.
+	v.children = append(v.children, node)
+	return nil
+}
+
+// orMeasurementTree determines if a tree (or subtree) represents a grouping of
+// exclusively measurement names OR'd together with EQ operators for the
+// measurements themselves. It collects the list of measurement names
+// encountered and records the validity of the tree.
+type orMeasurementTree struct {
+	measurementKey   string
+	measurementNames []string
+	valid            bool
+}
+
+func (v *orMeasurementTree) Visit(node influxql.Node) influxql.Visitor {
+	// A measurement occuring in a tree which has been proven to be invalid
+	// results in the measurement optimization not applying, so the evaluation can
+	// be stopped early in this case.
+	if !v.valid && len(v.measurementNames) > 0 {
+		return nil
+	}
+
+	switch n := node.(type) {
+	case *influxql.BinaryExpr:
+		if n.Op == influxql.OR {
+			return v
+		}
+
+		if name, ok := measurementNameFromBinary(n, v.measurementKey); ok {
```

Comment:
Do we have a test for something this misses? e.g.

```
(r._name = "foo" || r._name == "bar") && (r._value == r._name)
```

Reply:
I made some updates that we discussed which should alleviate this concern. Since an expression like `(_name = 'm0' OR _name = 'm1' OR _name = 'm2') AND (tag1 != 'foo' OR _name = 'm1')` should be able to apply the optimization (we know that the values must come from `m0`, `m1`, or `m2`, even though `m1` occurs on the left and right side of the AND), I updated the code to only look for a single group of exclusive OR'd measurements, and not invalidate the optimization if a measurement occurs elsewhere.

This simplifies looking for measurements anywhere else in the query, since it basically doesn't matter. I renamed this funciton to `measurementNameFromEqBinary` and added clarification that it only returns measurement names for EQ operations in the from of `_measurement = name`. For an EQ that has anything else, the tree will be invalidated.